README.md
microcom
========

microcom is a minimalistic terminal program for communicating with devices over
a serial connection (e.g. embedded systems, switches, modems). It features
connection via RS232 serial interfaces (including setting of transfer rates) as
well as in "Telnet mode" as specified in [RFC 2217].

[RFC 2217]: https://tools.ietf.org/html/rfc2217


Installation
------------

microcom depends on the [readline] library.

If you just cloned this repository, you also need to install [autoconf],
[automake], and the [autoconf archive] first. Then change to the project's root
directy and do:

```
autoreconf -i
```

If you extracted minicom from a release tarball, this previous step should not
be needed.

Now continue with building and installing minicom:

```
./configure
make
sudo make install
```

By default, minicom is installed into `/usr/local/bin/`. Use `./configure
--prefix=YOURPATH` to change that, and see `./configure --help` for more
options related to building and installation.

[readline]: https://tiswww.case.edu/php/chet/readline/rltop.html
[autoconf]: https://www.gnu.org/software/autoconf/
[automake]: https://www.gnu.org/software/automake/
[autoconf archive]: https://www.gnu.org/software/autoconf-archive/


Usage
-----

The typical usage with TTY devices looks like this:

```
microcom --speed=115200 --port=/dev/ttyS0
```

To connect to remote serial ports via RFC 2217, use the ``--telnet`` option instead:

```
microcom --speed=115200 --telnet=somehost:port
```

For the full list of options, see `microcom --help`.

During the connection, you can get to the microcom menu by pressing `Ctrl-\`.
Various options are available there, like setting flow  control, RTS and DTR.
See ``help`` for a full list.


License and Contributing
------------------------

microcom is free software and distributable under the GNU General Public
License, version 2. See the file `COPYING` in this repository for more
information.

Changes to microcom must be certified to be compatible with this license. For
this purpose, we use the Developer's Certificate of Origin 1.1; see the file
`DCO` in this repository.  If you can certify that the DCO applies for your
changes, add a line like the following:

```
Signed-off-by: Random J Developer <random@developer.example.org>
```

… containing your real name and e-mail address at the end of the patch
description (Git can do this for you when you use `git commit -s`).
Then send your patches to <oss-tools@pengutronix.de>.


ChangeLog
2014-09-22  Uwe Kleine-König <u.kleine-koenig@pengutronix.de>

  * Makefile: release 2014.09.0

2014-09-11  Uwe Kleine-König <u.kleine-koenig@pengutronix.de>

  * telnet.c: make telnet mode address family agnostic  Use getaddrinfo et al
  over gethostbyname. This allows (depending on system configuration) to use
  ipv6 connections. As an added bonus allow specifying ipv6 addresses in the
  usual square bracketed syntax.

2014-02-18  Sascha Hauer <s.hauer@pengutronix.de>

  * mux.c: Only send answerback string if one is set

2014-02-12  Sascha Hauer <s.hauer@pengutronix.de>

  * mux.c: factor out handle_receive_buf function  Instead of handling special
  characters in two places.

  * microcom.c, microcom.h, mux.c: Add ENQ/answerback support  This adds
  support for a configurable answerback string for an enquiry character (ENQ,
  ASCII 0x5). The answerback string can be specified using the --answerback
  option. If this option is not given, no answerback string is sent which is
  the behaviour without this patch.

  * microcom.c: update help text  Document the long options and some missing
  options.

2014-02-11  Sascha Hauer <s.hauer@pengutronix.de>

  * mux.c: remove unnecessary return

2014-02-11  Uwe Kleine-König <u.kleine-koenig@pengutronix.de>

  * ChangeLog, Makefile: release 2014.02.0

2014-02-11  Marc Kleine-Budde <mkl@pengutronix.de>

  * microcom.c: microcom: add more baudrates  This patch adds support for more
  baudrates.  Acked-by: Sascha Hauer <s.hauer@pengutronix.de>

  * commands.c, microcom.c, microcom.h: microcom: fix check if
  baudrate_to_flag() fails  baudrate_to_flag() returns a "speed_t" and -1 in
  case of an error. As speed_t is usually unsinged int an error is never
  detected.  This patch converts baudrate_to_flag() to pass speed_t via
  reference, so that the return value can be used to indicate an error. 
  Acked-by: Sascha Hauer <s.hauer@pengutronix.de>

2014-02-11  Sascha Hauer <s.hauer@pengutronix.de>

  * microcom.c: check mux_loop return value, exit with error  If mux_loop
  returns with an error report it and exit microcom with an error code.

  * microcom.h, mux.c: let mux_loop return an error  Instead of void.

  * mux.c: use -1 as invalid filedescriptor  0 is a valid filedescriptor, so
  don't use it to indicate a logfile not being opened.

2014-02-06  Sascha Hauer <s.hauer@pengutronix.de>

  * microcom.c, microcom.h, mux.c: Add support for listononly option  With the
  listenonly option microcom will only listen on the port, but never send
  anything back. Also the local terminal is not modified. This is useful for
  noninteractively creating logfiles from remote ports.

  * commands.c, microcom.c, microcom.h, mux.c: Add logfile support  logfile
  support was lost during restructuring earlier. Add it again.

2014-01-13  Uwe Kleine-König <u.kleine-koenig@pengutronix.de>

  * ChangeLog, Makefile: release 2014.01.1

  * microcom.c: Fix compiler warnings and segfault  Commit 3337bd893b6b dropped
  the comma at the end of the format string by mistake resulting in several
  compiler warnings:  microcom.c:150:3: warning: format ‘%s’ expects
  argument of type ‘char *’, but argument 3 has type ‘int’ [-Wformat]
  microcom.c:150:3: warning: format ‘%d’ expects argument of type
  ‘int’, but argument 4 has type ‘char *’ [-Wformat] microcom.c:150:3:
  warning: format ‘%s’ expects argument of type ‘char *’, but argument
  5 has type ‘int’ [-Wformat] microcom.c:150:3: warning: format ‘%x’
  expects a matching ‘unsigned int’ argument [-Wformat]  and a segfault
  when main_usage is called. So readd that comma.  Fixes: 3337bd893b6b (fix
  help string to include -d and -f)

  * .gitignore: .gitignore: don't ignore .gitignore, anchor microcom entry

  * commands.c, commands_fsl_imx.c, parser.c: Add explicit copyright and
  license to three files  These files were only touched by Sascha and Marc back
  in 2010.  Acked-by: Marc Kleine-Budde <mkl@pengutronix.de> Acked-by: Sascha
  Hauer <s.hauer@pengutronix.de>

2014-01-10  Uwe Kleine-König <uwe@kleine-koenig.org>

  * ChangeLog, Makefile: release 2014.01.0

  * gitlog2changelog.py: gitlog2changelog: adapt to output generated by newer
  git  ... that's why you shouldn't parse output of porcelain tools

2014-01-04  Uwe Kleine-König <uwe@kleine-koenig.org>

  * microcom.c: fix help string to include -d and -f  Also drop wrong comment
  about session logging

2013-12-30  Uwe Kleine-König <uwe@kleine-koenig.org>

  * microcom.1: add a simple manpage  The text bases on work done by Alexander
  Reichle-Schmehl for Debian, licensed under GPL v2+. I converted it from sgml
  to plain roff and fixed several things.  Closes:
  http://bugs.debian.org/698436

2013-06-10  Markus Pargmann <mpa@pengutronix.de>

  * Makefile: Make: Fix default link rule for --as-needed  microcom does not
  compile with LDFLAGS="-Wl,--as-needed" because of a default make rule with
  LDFLAGS before all object files. The linker seems to parse the commandline
  only once, which results in this case in undefined references.  This patch
  defines a custom link rule with LDFLAGS at the end of the command line.

2013-06-10  Michael Olbrich <m.olbrich@pengutronix.de>

  * microcom.c: add signal handler for SIGQUIT  it seems readline() send
  SIGQUIT on ctrl-\

2013-04-29  Uwe Kleine-König <u.kleine-koenig@pengutronix.de>

  * mux.c: Fix a few typos in comments and point out a layer violation

  * microcom.c: fix newline handling  Disable any nl <-> cr translation. This
  makes it possible to drive a debian installer and still directing barebox and
  Linux continues to work.  While at it drop cleaning BRKINT as it doesn't do
  anything with IGNBRK set.

2012-06-14  Marc Kleine-Budde <mkl@pengutronix.de>

  * ChangeLog, Makefile: release 2012.06.0

2012-02-18  Jonas Zetterberg <jozz@jozz.se>

  * serial.c: Remove lockfile on open serial failure  Since the lockfile is
  created before the serial device is opened, if the serial device can not be

microcom.c
/******************************************************************
** File: microcom.c
** Description: the main file for microcom project
**
** Copyright (C)1999 Anca and Lucian Jurubita <ljurubita@hotmail.com>.
** All rights reserved.
****************************************************************************
** This program is free software; you can redistribute it and/or
** modify it under the terms of the GNU General Public License
** as published by the Free Software Foundation; either version 2
** of the License, or (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details at www.gnu.org
****************************************************************************
** Rev. 1.0 - Feb. 2000
** Rev. 1.01 - March 2000
** Rev. 1.02 - June 2000
****************************************************************************/
#define _GNU_SOURCE
#include "microcom.h"

#include <unistd.h>
#include <getopt.h>
#include <sys/socket.h>
#include <string.h>
#include <stdint.h>
#include <sys/types.h>
#include <sys/stat.h>

static struct termios sots;	/* old stdout/in termios settings to restore */

struct ios_ops *ios;
int debug;

void init_terminal(void)
{
	struct termios sts;

	memcpy(&sts, &sots, sizeof (sots));     /* to be used upon exit */

	/* again, some arbitrary things */
	sts.c_iflag &= ~(IGNCR | INLCR | ICRNL);
	sts.c_iflag |= IGNBRK;
	sts.c_lflag &= ~ISIG;
	sts.c_cc[VMIN] = 1;
	sts.c_cc[VTIME] = 0;
	sts.c_lflag &= ~ICANON;
	/* no local echo: allow the other end to do the echoing */
	sts.c_lflag &= ~(ECHO | ECHOCTL | ECHONL);

	tcsetattr(STDIN_FILENO, TCSANOW, &sts);
}

void restore_terminal(void)
{
	tcsetattr(STDIN_FILENO, TCSANOW, &sots);
}

int baudrate_to_flag(int speed, speed_t *flag)
{
	switch(speed) {
	case 50: *flag = B50; return 0;
	case 75: *flag = B75; return 0;
	case 110: *flag = B110; return 0;
	case 134: *flag = B134; return 0;
	case 150: *flag = B150; return 0;
	case 200: *flag = B200; return 0;
	case 300: *flag = B300; return 0;
	case 600: *flag = B600; return 0;
	case 1200: *flag = B1200; return 0;
	case 1800: *flag = B1800; return 0;
	case 2400: *flag = B2400; return 0;
	case 4800: *flag = B4800; return 0;
	case 9600: *flag = B9600; return 0;
	case 19200: *flag = B19200; return 0;
	case 38400: *flag = B38400; return 0;
	case 57600: *flag = B57600; return 0;
	case 115200: *flag = B115200; return 0;
	case 230400: *flag = B230400; return 0;
#ifdef B460800
	case 460800: *flag = B460800; return 0;
#endif
#ifdef B500000
	case 500000: *flag = B500000; return 0;
#endif
#ifdef B576000
	case 576000: *flag = B576000; return 0;
#endif
#ifdef B921600
	case 921600: *flag = B921600; return 0;
#endif
#ifdef B1000000
	case 1000000: *flag = B1000000; return 0;
#endif
#ifdef B1152000
	case 1152000: *flag = B1152000; return 0;
#endif
#ifdef B1500000
	case 1500000: *flag = B1500000; return 0;
#endif
#ifdef B2000000
	case 2000000: *flag = B2000000; return 0;
#endif
#ifdef B2500000
	case 2500000: *flag = B2500000; return 0;
#endif
#ifdef B3000000
	case 3000000: *flag = B3000000; return 0;
#endif
#ifdef B3500000
	case 3500000: *flag = B3500000; return 0;
#endif
#ifdef B4000000
	case 4000000: *flag = B4000000; return 0;
#endif
	default:
		printf("unknown speed: %d\n",speed);
		return -1;
	}
}

int flag_to_baudrate(speed_t speed)
{
	switch(speed) {
	case B50: return 50;
	case B75: return 75;
	case B110: return 110;
	case B134: return 134;
	case B150: return 150;
	case B200: return 200;
	case B300: return 300;
	case B600: return 600;
	case B1200: return 1200;
	case B1800: return 1800;
	case B2400: return 2400;
	case B4800: return 4800;
	case B9600: return 9600;
	case B19200: return 19200;
	case B38400: return 38400;
	case B57600: return 57600;
	case B115200: return 115200;
	case B230400: return 230400;
	default:
		printf("unknown speed: %d\n",speed);
		return -1;
	}
}

void microcom_exit(int signal)
{
	printf("exiting\n");

	ios->exit(ios);
	tcsetattr(STDIN_FILENO, TCSANOW, &sots);

	exit(0);
}

/********************************************************************
 Main functions
 ********************************************************************
 static void help_usage(int exitcode, char *error, char *addl)
      help with running the program
      - exitcode - to be returned when the program is ended
      - error - error string to be printed
      - addl - another error string to be printed
 static void cleanup_termios(int signal)
      signal handler to restore terminal set befor exit
 int main(int argc, char *argv[]) -
      main program function
********************************************************************/
void main_usage(int exitcode, char *str, char *dev)
{
	fprintf(stderr, "Usage: microcom [options]\n"
		" [options] include:\n"
		"    -p, --port=<devfile>                 use the specified serial port device (%s);\n"
		"    -s, --speed=<speed>                  use specified baudrate (%d)\n"
		"    -t, --telnet=<host:port>             work in telnet (rfc2217) mode\n"
		"    -c, --can=<interface:rx_id:tx_id>    work in CAN mode\n"
		"                                         default: (%s:%x:%x)\n"
		"    -f, --force                          ignore existing lock file\n"
		"    -d, --debug                          output debugging info\n"
		"    -l, --logfile=<logfile>              log output to <logfile>\n"
		"    -o, --listenonly                     Do not modify local terminal, do not send input\n"
		"                                         from stdin\n"
		"    -a,  --answerback=<str>              specify the answerback string sent as response to\n"
		"                                         an ENQ (ASCII 0x05) Character\n"
		"    -h, --help                           This help\n",
		DEFAULT_DEVICE, DEFAULT_BAUDRATE,
		DEFAULT_CAN_INTERFACE, DEFAULT_CAN_ID, DEFAULT_CAN_ID);
	fprintf(stderr, "Exitcode %d - %s %s\n\n", exitcode, str, dev);
	exit(exitcode);
}

int opt_force = 0;
int current_speed = DEFAULT_BAUDRATE;
int current_flow = FLOW_NONE;
int listenonly = 0;

int main(int argc, char *argv[])
{
	struct sigaction sact;  /* used to initialize the signal handler */
	int opt, ret;
	char *hostport = NULL;
	int telnet = 0, can = 0;
	char *interfaceid = NULL;
	char *device = DEFAULT_DEVICE;
	char *logfile = NULL;
	speed_t flag;

	struct option long_options[] = {
		{ "help", no_argument, 0, 'h' },
		{ "port", required_argument, 0, 'p'},
		{ "speed", required_argument, 0, 's'},
		{ "telnet", required_argument, 0, 't'},
		{ "can", required_argument, 0, 'c'},
		{ "debug", no_argument, 0, 'd' },
		{ "force", no_argument, 0, 'f' },
		{ "logfile", required_argument, 0, 'l'},
		{ "listenonly", no_argument, 0, 'o'},
		{ "answerback", required_argument, 0, 'a'},
		{ 0, 0, 0, 0},
	};

	while ((opt = getopt_long(argc, argv, "hp:s:t:c:dfl:oi:a:", long_options, NULL)) != -1) {
		switch (opt) {
			case 'h':
			case '?':
				main_usage(1, "", "");
				exit(0);
			case 'p':
				device = optarg;
				break;
			case 's':
				current_speed = strtoul(optarg, NULL, 0);
				break;
			case 't':
				telnet = 1;
				hostport = optarg;
				break;
			case 'c':
				can = 1;
				interfaceid = optarg;
				break;
			case 'f':
				opt_force = 1;
				break;
			case 'd':
				debug = 1;
				break;
			case 'l':
				logfile = optarg;
				break;
			case 'o':
				listenonly = 1;
				break;
			case 'a':
				answerback = optarg;
				break;
		}
	}

	if (answerback) {
		ret = asprintf(&answerback, "%s\n", answerback);
		if (ret < 0)
			exit (1);
	}

	commands_init();
	commands_fsl_imx_init();

	if (telnet && can)
		main_usage(1, "", "");

	if (telnet)
		ios = telnet_init(hostport);
	else if (can)
		ios = can_init(interfaceid);
	else
		ios = serial_init(device);

	if (!ios)
		exit(1);

	if (logfile) {
		ret = logfile_open(logfile);
		if (ret < 0)
			exit(1);
	}

	ret = baudrate_to_flag(current_speed, &flag);
	if (ret)
		exit(1);

	current_flow = FLOW_NONE;
	ios->set_speed(ios, flag);
	ios->set_flow(ios, current_flow);

	if (!listenonly) {
		printf("Escape character: Ctrl-\\\n");
		printf("Type the escape character followed by c to get to the menu or q to quit\n");

		/* Now deal with the local terminal side */
		tcgetattr(STDIN_FILENO, &sots);
		init_terminal();

		/* set the signal handler to restore the old
		 * termios handler */
		sact.sa_handler = &microcom_exit;
		sigaction(SIGHUP, &sact, NULL);
		sigaction(SIGINT, &sact, NULL);
		sigaction(SIGPIPE, &sact, NULL);
		sigaction(SIGTERM, &sact, NULL);
		sigaction(SIGQUIT, &sact, NULL);
	}

	/* run the main program loop */
	ret = mux_loop(ios);
	if (ret)
		fprintf(stderr, "%s\n", strerror(-ret));

	ios->exit(ios);

	if (!listenonly)
		tcsetattr(STDIN_FILENO, TCSANOW, &sots);

	exit(ret ? 1 : 0);
}
